### PR TITLE
1366 batch change age group and enrollment

### PR DESCRIPTION
**Pull request by nickbristow**
## Background
Adds the ability to change enrollment and age group in batches from roster page

## GitHub Issue
#1366 

## Validation Plan
Please outline the steps you've already taken to validate this body of work, along with any additional steps that need to be taken in order to complete testing.

## Automated Testing
- [ ] All unit tests are passing.
- [ ] If applicable, unit tests have been added to cover this changeset.

## Additional Context
If there's anything additional information that those reviewing or validating this work should be aware of, please specify it here.

### CODE REVIEW

File: client/src/containers/EditRecord/EnrollmentFunding/ChangeEnrollment/Form.tsx [hunk 1]
```diff
@@ -32,25 +34,43 @@ export const ChangeEnrollmentForm: React.FC<ChangeEnrollmentFormProps> = ({
...
-    (site) => site.organizationId === child.organization?.id
-  );
-
   const { accessToken } = useContext(AuthenticationContext);
-
   const [loading, setLoading] = useState(false);
+
+  if (!child && !orgId) return <></>;
+
+  const organizationId = child?.organization?.id || orgId || 1;
```
Comment (bamader) [inline client/src/containers/EditRecord/EnrollmentFunding/ChangeEnrollment/Form.tsx:46]: Curious on the ordering here? If we're explicitly supplying an `orgId` wouldn't we want that to have first priority in the event the `OR` short-circuits early?

Comment (bamader) [reply client/src/containers/EditRecord/EnrollmentFunding/ChangeEnrollment/Form.tsx:46]: Oh unless the `orgId` parameter is only used in the event of batch changing? If that is the case, can we change some variable names or add some documentation describing the use case? It was not at all obvious to me on reading the function signature that we wouldn't be using all of the parameters on each call. This is especially true for `selectedIds`, which could refer to children, sites, etc.

File: client/src/containers/Roster/Roster.scss [hunk 2]
```diff
@@ -71,3 +71,7 @@
 .bx--table-sort.bx--table-sort--hover {
   background-color: #eddbe9;
 }
+
+.flatpickr-calendar {
+  z-index: 99999999999 !important;
+}
```
Comment (bamader) [inline client/src/containers/Roster/Roster.scss:77]: wat

**Review by bamader — APPROVED** [review]
Seems legit

File: client/src/containers/EditRecord/EnrollmentFunding/ChangeEnrollment/Form.tsx [hunk 3]
```diff
@@ -32,25 +34,47 @@ export const ChangeEnrollmentForm: React.FC<ChangeEnrollmentFormProps> = ({
...
-    apiPost(`children/${child.id}/change-enrollment`, updatedData, {
-      accessToken,
-      jsonParse: false,
-    })
-      .then(afterSaveSuccess)
-      .catch(afterSaveFailure)
-      .finally(() => setLoading(false));
+    console.log('selectedChildren', selectedChildren);
+    console.log('selectedChildren.keys', Object.keys(selectedChildren));
+    if (child)
```
[the remark this answers is not in the document]
Comment (nickbristow) [reply client/src/containers/EditRecord/EnrollmentFunding/ChangeEnrollment/Form.tsx:54]: 👍

File: client/src/containers/Roster/Roster.scss (hunk 2 above)
Comment (nickbristow) [reply client/src/containers/Roster/Roster.scss:77]: need to make sure the calendar is on top.

File: client/src/containers/EditRecord/EnrollmentFunding/ChangeEnrollment/Form.tsx (hunk 1 above)
Comment (nickbristow) [reply client/src/containers/EditRecord/EnrollmentFunding/ChangeEnrollment/Form.tsx]: @bamader this probably makes more sense, i was just trying to get typescript to stop throwing an error. Now its this

```
const organizationId = orgId || child?.organization?.id;
  if (!organizationId) return <></>;
```
If we don't have an organizationId, we can't do anything, so we just don't return a form. This is really just to keep it from breaking.

**Review by GordonFarrell — APPROVED** [review]

File: client/src/containers/Roster/Roster.tsx [hunk 4]
```diff
@@ -94,6 +100,7 @@ const childToRow = (child: RosterChild) => {
     }),
     site: site?.siteName,
     missingInfo: !!child.validationErrors?.length,
+    child: child,
```
Comment (JordanGuinn) [inline client/src/containers/Roster/Roster.tsx:103]: I don't follow - why does the child have an instance of itself as a parameter?

File: client/src/containers/Roster/BatchChangeEnrollment.tsx [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+
+type ChildIdName = {
+  id: string;
+  name: string;
+};
+
+export type BatchChangeEnrollmentProps = {
+  icon?: any;
+  orgId: number;
+  selectedChildren: any;
```
Comment (JordanGuinn) [inline client/src/containers/Roster/BatchChangeEnrollment.tsx:19]: We should definitely give this an explicit type, even if it's just `Partial<ChildRow>` or something?

File: client/src/containers/Roster/BatchChangeEnrollment.tsx [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+  setAlerts: Function;
+  fetchRoster: Function;
+  tabIndex: number;
+} & Omit<
+  ChangeEnrollmentFormProps,
+  'afterSaveFailure' | 'topHeadingLevel' | 'afterSaveSuccess'
+>;
+
+export const BatchChangeEnrollment: React.FC<BatchChangeEnrollmentProps> = ({
+  child,
```
Comment (JordanGuinn) [inline client/src/containers/Roster/BatchChangeEnrollment.tsx:29]: Why does this accept both `child` **and** `selectedChildren` as potential parameters?  I know that it exists on `ChangeEnrollmentFormProps`, but it probably should be omitted on `BatchChangeEnrollmentProps`

File: client/src/containers/EditRecord/EnrollmentFunding/ChangeEnrollment/Form.tsx [hunk 7]
```diff
@@ -32,25 +34,45 @@ export const ChangeEnrollmentForm: React.FC<ChangeEnrollmentFormProps> = ({
...
+      apiPost(`children/${child.id}/change-enrollment`, updatedData, {
+        accessToken,
+        jsonParse: false,
+      })
+        .then(afterSaveSuccess)
+        .catch(afterSaveFailure)
+        .finally(() => setLoading(false));
+    if (selectedChildren) {
+      console.log('selectedChildren', selectedChildren);
```
Comment (JordanGuinn) [inline client/src/containers/EditRecord/EnrollmentFunding/ChangeEnrollment/Form.tsx:63]: nit: remove `console.log`

File: client/src/containers/Roster/BatchChangeEnrollment.tsx (hunk 6 above)
Comment (nickbristow) [reply client/src/containers/Roster/BatchChangeEnrollment.tsx:29]: yes, left over from where i started when i created this component.

File: client/src/containers/Roster/Roster.tsx (hunk 4 above)
Comment (nickbristow) [reply client/src/containers/Roster/Roster.tsx]: I was trying to debug something, and left this in. Removed

File: client/src/containers/Roster/BatchChangeEnrollment.tsx [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+import { Modal } from '@ctoec/component-library';
+import pluralize from 'pluralize';
+import { TableBatchAction } from 'carbon-components-react';
+import {
+  ChangeEnrollmentForm,
+  ChangeEnrollmentFormProps,
+} from '../EditRecord/EnrollmentFunding/ChangeEnrollment/Form';
+
+export type idNameType = {
+  [key: string]: string;
```
Comment (JordanGuinn) [inline client/src/containers/Roster/BatchChangeEnrollment.tsx:12]: What are the actual fields that we care about here?  Can we just enumerate those?  Like if it ends up being just `id` and `name`, can this be something like:

`export type EnrolledChild = Pick<ChildRow, 'id' | 'name'>;`

**Review by JordanGuinn — APPROVED** [review]
LGTM - just one final q about a particular TS type

File: client/src/containers/Roster/BatchChangeEnrollment.tsx (hunk 8 above)
Comment (JordanGuinn) [reply client/src/containers/Roster/BatchChangeEnrollment.tsx:12]: Discussed async in Slack - type makes sense in it's current state, now just requesting a naming tweak to make it more verbose

**Comment by bamader** [comment]
## Validation Plan

- [ ] `change age group` button only appears when # of selected children is >= 1
- [ ] Modal appears on button click, with all mentions of `enrollment` changed to `enrollments`
- [ ] Can successfully batch change single child's age group
- [ ] Can successfully batch change group of children's age groups
- [ ] Success alert shows how many children were updated
- [ ] Can batch change some children's age groups and get notifications of children that failed
- [ ] Can successfully batch single child's enrollment
- [ ] Can successfully batch change group of children's enrollments 
- [ ] Success alert shows how many children were updated
- [ ] Can batch change some children's enrollments and get notifications of children that failed

**Comment by bamader** [comment]
Nick's addressed the outstanding issues here, so that completes validations. Once we get/if we already have bug tickets open for the stuff we talked about in the huddle, this PR is ready to go!

**Pull request merged**